src/charm.py

```diff
@@ -126,43 +131,50 @@ def _on_cluster_relation_changed(self, event: EventBase) -> None:
             return
 
         # check whether restart is needed for all `*_changed` events
-        self.on[f"{self.restart.name}"].acquire_lock.emit()
+        # only restart where necessary to avoid slowdowns
+        if (self.config_changed() or self.tls.upgrading) and self.cluster.started:
+            self.on[f"{self.restart.name}"].acquire_lock.emit()
 
         # ensures events aren't lost during an upgrade on single units
         if self.tls.upgrading and len(self.cluster.peer_units) == 1:
             event.defer()
 
-    def _restart(self, _) -> None:
+    def _restart(self, event: EventBase) -> None:
         """Handler for emitted restart events."""
-        # 'snap restart' starts the service, so this can cause issues if ran before `init_server()`
-        if not self.cluster.started:
+        if not self.cluster.stable:
+            logger.debug("restart - cluster not stable - deferring")
+            event.defer()
             return
 
-        if self.config_changed() or self.cluster.manual_restart:
-            logger.info(f"Server.{self.cluster.get_unit_id(self.unit)} restarting")
-            self.snap.restart_snap_service(snap_service="daemon")
+        logger.info(f"Server.{self.cluster.get_unit_id(self.unit)} restarting")
+        self.snap.restart_snap_service(snap_service="daemon")
 
-            # gives time for server to rejoin quorum, as command exits too fast
-            # without, other units might restart before this unit rejoins, losing quorum
-            time.sleep(5)
+        # gives time for server to rejoin quorum, as command exits too fast
+        # without, other units might restart before this unit rejoins, losing quorum
+        time.sleep(5)
 
-            self.unit.status = ActiveStatus()
+        self.unit.status = ActiveStatus()
 
         # Indicate that unit has completed restart on password rotation
         if self.cluster.relation.data[self.app].get("rotate-passwords"):
             self.cluster.relation.data[self.unit]["password-rotated"] = "true"
 
+        if not self.tls.unit_unified(unit=self.unit) and self.tls.upgrading:
+            # if unit restarted with `upgrading` app data, must be unified
+            logger.debug(f"{self.unit.name} setting unified status")
```

Comment:
**question**  Looks a bit weird to have a if statement JUST for a log...but if this is useful I'm of course not against. Just it seems a sort of leftover, so I thought it was worth flagging this

Reply:
You're right, it was for debugging. 
Although useful, it's a bit bloaty to have 3 lines for it, will remove!